pilot/pkg/model/network.go

```diff
@@ -544,15 +561,31 @@ func newClient() (*dnsClient, error) {
 	return c, nil
 }
 
+// for more informative logging of dns errors
+func getReqNames(req *dns.Msg) []string {
+	names := make([]string, 0, 1)
+	for _, qq := range req.Question {
+		names = append(names, qq.Name)
+	}
+	return names
+}
+
 func (c *dnsClient) Query(req *dns.Msg) *dns.Msg {
 	var response *dns.Msg
 	for _, upstream := range c.resolvConfServers {
 		cResponse, _, err := c.Exchange(req, upstream)
 		if err == nil {
 			response = cResponse
-			break
+			code := response.MsgHdr.Rcode
+			if code == dns.RcodeSuccess {
+				break
+			}
+			codeString := dns.RcodeToString[code]
+			// TODO make Debug as to not spam the logs on retries?
+			log.Infof("upstream dns error: %v: %v: %v", upstream, getReqNames(req), codeString)
```

Comment:
Is there an error when only one type of DNS record supported?

Reply:
Well-behaved DNS servers should return NOERROR with an empty record set, as per RFC, but [that's not guaranteed](https://datatracker.ietf.org/doc/html/rfc4074#section-3). If an upstream DNS server has no proper IPv6 support then each AAAA lookup can lead to a logged NXDOMAIN error.

I think we can omit logging here if the status is NXDOMAIN. If there is either A or AAAA record the error for another one is redundant, and if both are missing the error will be logged in the downstream function NetworkManager.resolveHostnameGateways():
https://github.com/istio/istio/blob/e67c34b8c005ec13d2a3b2dd897b45124a02a5ca/pilot/pkg/model/network.go#L191-L193

However if there is a concern for other unexpected errors from the upstream I think it should be indeed downgraded to `log.Debugf`.

Reply:
Yes, that's my concern

Reply:
Changed to `log.Debugf`. Not skipping on NXDOMAIN then as it may indicate a misbehaving upstream and it's on the debug log level now.